benchmark/server/start.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2021 PairMesh, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"syscall"
+
+	"go.uber.org/zap"
+)
+
+const bufferSize = 512
+
+// Start function starts a simple echo server as server side of pairbenchmark
+// If cfg.isBounce is true, it echoes whatever it hears from incoming connection
+// Otherwise, it always echoes "OK" as minimal backward payload
+func Start(cfg *Config) error {
+	zap.L().Info("Starting pairbenchmark server")
+	addr := fmt.Sprintf("%s:%d", "0.0.0.0", cfg.Port())
+	s, err := net.Listen("tcp", addr)
+	if err != nil {
+		zap.L().Error("Error setting up server")
+		return err
+	}
+
+	zap.L().Info(fmt.Sprintf("Started pairbenchmark server on port %d", cfg.Port()))
+
+	defer s.Close()
```

Comment:
Will you replace the generic server to our relay server? I think the benchmark doesn't reflect any bottleneck of our system.

Reply:
The whole points of having a generic echo server is that our users might not care about our internals. They don't care relay or portal or whatever. They just want to see the end-to-end performance from one device to another device. This is why the server is a generic one and is supposed to be deployed as another node to provide end-to-end perf results, rather than as a relay server.

In this case the performance is due to multiple factors:
- Local network protocol stack IO and computation (generic)
- TUN IO (PairMesh)
- Encryption and Decryption (PairMesh)
- Actual network packet transmission (generic)
- Relay in some cases (PairMesh)

They sum up to be what the results reflect.

But I totally understand your point that from development perspective, we need to break the end-to-end scenario down to finer grain and more understandable cases. I'll add another mode to run server as relay server.

Reply:
Make sense, We must to provide some documentations for users.

Reply:
Make sense, We must to provide some documentations for users.